File: Java/com/leetcode/searchA2DMatrix/SearchA2DMatrix.java
package com.leetcode.searchA2DMatrix;

public class SearchA2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        // Binary Search
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0 || target < matrix[0][0] || target > matrix[matrix.length - 1][matrix[0].length - 1]) {
            return false;
        }
        
        int rows = matrix.length,
            columns = matrix[0].length,
            start = 0,
            end = rows - 1,
            mid = -1;
        
        while (start <= end) {
            mid = (end + start) / 2;
            if (target < matrix[mid][0]) {
                end = mid - 1;
            } else if (target > matrix[mid][columns - 1]){
                start = mid + 1;
            } else if (target > matrix[mid][0] && target < matrix[mid][columns - 1]) {
                break;
            } else if (target == matrix[mid][0] || target == matrix[mid][columns - 1]) {
                return true;
            }
        }
        
        
         
        // Using another binary search
        start = 0;
        end = columns - 1;
        int rowIndex = mid;
        mid = -1;
        
        while (start <= end) {
            mid = (end + start) / 2;
            if (target < matrix[rowIndex][mid]) {
                end = mid - 1;
            } else if (target == matrix[rowIndex][mid]) {
                return true;
            } else {
                start = mid + 1;
            }
        }
        
        return false;
        

        
    }
}

File: Python/search_a_2d_matrix/search_a_2d_matrix.py
class Solution(object):
	def searchMatrix(self, matrix, target):
		idx_row = self.search_row(0, len(matrix)-1, matrix, target)
		return self.search_col(0, len(matrix[idx_row])-1, matrix[idx_row], target)

	def search_row(self, start, end, matrix, target):
		if start == end:
			if matrix[start][0] > target:
				return start - 1
			else:
				return start
		median = (start + end) // 2
		if matrix[median][0] == target:
			return median
		if matrix[median][0] > target:
			return self.search_row(start, median, matrix, target)
		else:
			return self.search_row(median+1, end, matrix, target)

	def search_col(self, start, end, row, target):
		if start == end:
			if row[start] != target:
				return False
			else:
				return True
		median = (start + end) // 2
		if row[median] == target:
			return True
		if row[median] > target:
			return self.search_col(start, median, row, target)
		else:
			return self.search_col(median+1, end, row, target)
		
		
